util/reggen/gen_rust.py

```diff
@@ -0,0 +1,394 @@
+from .params import LocalParam
+from .register import Register
+from .multi_register import MultiRegister
+from .signal import Signal
+from .window import Window
+
+
+def genout(outfile: TextIO, msg: str) -> None:
+    outfile.write(msg)
+
+def to_snake_case(s: str) -> str:
+    val = []
+    for i, ch in enumerate(s):
+      if i > 0 and ch.isupper():
+        val.append('_')
+      val.append(ch)
+    return ''.join(val)
+
+def as_define(s: str) -> str:
+    s = s.upper()
+    r = ''
+    for i in range(0, len(s)):
+        r += s[i] if s[i].isalnum() else '_'
+    return r
+
+
+def first_line(s: str) -> str:
+    """Returns the first line of a multi-line string"""
+    return s.splitlines()[0]
+
+
+def format_comment(s: str) -> str:
+    """Formats a string to comment wrapped to an 80 character line width
+
+    Returns wrapped string including newline and // comment characters.
+    """
+    return '\n'.join(
+        textwrap.wrap(
+            s, width=77, initial_indent='// ', subsequent_indent='// ')) + '\n'
+
+
+def is_negative(x: str) -> bool:
+    """ Checks if a decimal string is a negative value. """
+
+    # Try and parse decimal string.
+    try:
+        n = int(x)
+        return n < 0
+    except ValueError:
+        return False
+
+
+def data_type(name: str, val: str) -> str:
+    """ Returns proper data type for name-value pair. """
+
+    if name.endswith("_OFFSET") or name.endswith("_BASE_ADDR"):
+        return "usize"
+    elif is_negative(val):
+        return "i32"
+    else:
+        return "u32"
+
+
+def gen_const(name: str,
+               val: str,
+               existing_defines: Set[str],
+               indent: str = '  ') -> str:
+    r"""Produces a pub const string. Result includes newline.
+
+    Arguments:
+    name - Name of the constant
+    val - Value of the constant
+    existing_defines - set of already generated define names.
+        Error if `name` is in `existing_defines`.
+    indent - Gives string to prepend on any new lines produced by
+    wrapping (default '  ')
+
+    Example result:
+    name = 'A_NAME'
+    val = '10'
+
+    pub const A_NAME: u32 = 10
+
+    """
+
+    if name in existing_defines:
+        log.error("Duplicate pub const for " + name)
+        sys.exit(1)
+
+    define_declare = 'pub const ' + name + ': ' + data_type(name, val)
+
+    oneline_define = define_declare + ' = ' + val + ';'
+
+    existing_defines.add(name)
+
+    return oneline_define + '\n'
+
+
+def gen_const_register(outstr: TextIO,
+                         reg: Register,
+                         comp: str,
+                         width: int,
+                         rnames: Set[str],
+                         existing_defines: Set[str]) -> None:
+    rname = reg.name
+    offset = reg.offset
+
+    genout(outstr, format_comment(first_line(reg.desc)))
+    defname = as_define(comp + '_' + rname)
+    genout(
+        outstr,
+        gen_const(defname + '_REG_OFFSET', hex(offset), existing_defines))
```

Comment:
The auto-formatter strikes again! :-) Maybe define a variable for the result of `gen_const` to tidy up the code. Similarly with the calls `gen_const` in the loop below.

Reply:
Done.